### ISSUE TITLE
Support yarn workspaces (monorepos)

### ISSUE THREAD

**Issue by Hypnosphi** [depcheck/depcheck#245]
Currently, depcheck only checks root level dependencies

**Comment by Zjaaspoer** [comment]
Yes! Right now repo's using Yarn's workspaces are uncheckable by depcheck

**Comment by Hypnosphi** [comment]
Well, they're not completely uncheckable, you just have to do it separately for each package

**Comment by Zjaaspoer** [comment]
Well it gives false positives for the dependencies used within those workspaces. Workaround could be to ignore the workspaces directory I guess :)

**Comment by peterschussheim** [comment]
Are there alternative 'dependency checkers' that work correctly with `yarn workspaces`?  Ignoring workspaces directory isn't a solution.

**Comment by mnkhouri** [comment]
@Hypnosphi @Zjaaspoer @peterschussheim I don't have much experience with yarn workspaces, would you mind describing the ideal experience you'd like to see in a project with yarn workspaces? Would it be:

1. the tool determines what workspaces exist, then runs and generates a report for each workspace?
2. the tool ignores all workspaces automatically and just checks the root?
3. something different?

**Comment by benjie** [comment]
In my opinion, ideally there'd be two modes.

Mode 1 (independent workspaces, e.g. to be published to npm): each individual workspace is expected to have its own deps specified in the workspace's own package.json (and the root package.json is basically ignored dependencies-wise)

Mode 2 (dependent workspaces, e.g. to be ran in the context of the monorepo): each individual workspaces is expected to have its own deps specified in the workspace's own package.json *OR* the root package.json

In both cases, a) dependencies in the workspace's own package.json that are not used within that workspace should be flagged as unnecessary, b) dependencies required by the workspace might actually be installed at the root level due to yarn hoisting, this is fine and shouldn't raise any issues.

Definitely option 1 of the options you list would be preferred. I think the algorithm would be something like this (on abort just fall back to old behaviour):

1. starting at the relevant package.json, OriginalPackageJson, from where `depcheck` was called
2. find the nearest `yarn.lock` file by looking in the current folder or traversing up the tree until you find one (abort if not found)
3. find the package.json associated with this yarn.lock, RootPackageJson, (i.e. in the same folder) (abort if not found)
4. look for the "workspaces" key in RootPackageJson (abort if not found)
5. check that either RootPackageJson === OriginalPackageJson or OriginalPackageJson is in one of the RootPackageJson workspaces (abort if not)
6. if RootPackageJson === OriginalPackageJson then do a report for all workspaces (and the root?); otherwise do a report for just the selected workspaces

Hope this helps!

**Comment by mnkhouri** [comment]
Thanks for the clear and detailed comment @benjie!

I think it would be straightforward to implement the "Mode 1 (independent workspaces)" that you described [by adding some new functionality to cli.js](https://github.com/depcheck/depcheck/blob/539c6dd/src/cli.js#L76-L127), the main entrypoint for the program. There, we could do the directory traversal / parsing of `yarn.lock` and `package.json` that you described, and run the tool on each workspace we found. Since we'll be building up a list of workspaces while doing that, it would be easy to also run the tool on the root of the project and automatically exclude those workspaces.

After that chunk of work is complete, "Mode 2 (dependent workspaces)" seems easy enough to add on top of it.

For anyone interested in giving this a go, PRs are very welcome!

**Comment by oreporan** [comment]
for a workaround, I'm using `yarn workspaces` with `lerna`
then I just gave it :
`lerna exec --no-bail -- depcheck  --json \>\> ./depcheck.json `

That created the file in each directory and let me work on fixing each package

**Comment by pushred** [comment]
Another Lerna workaround (approach probably works for Workspaces as well):

- Create a npm script in each package that points to a root-level depcheck install, e.g.

```json
"scripts": {
  "depcheck": "../../node_modules/.bin/depcheck ."
}
```

- `npx lerna run depcheck --no-bail`

The output is separated by package, unlike `exec`

```
> api@1.0.0 depcheck /Users/example/monorepo/packages/api
> ../../node_modules/.bin/depcheck .

Missing dependencies
* lodash

>ui@1.0.0 depcheck /Users/example/monorepo/packages/ui
> ../../node_modules/.bin/depcheck .

Missing dependencies
* axios
* lodash
```

**Comment by sveyret** [comment]
Could the PR #437 solve your problems?

**Comment by sveyret** [comment]
I think that PR #473 now delivered in release 0.9.1 is correcting this issue.
@rumpl, this issue can probably be closed.

**Comment by pushred** [comment]
I'm still experiencing issues with root-installed dependencies: #509

**Comment by kingdaro** [comment]
Just here to unstale this issue, would really love to have this feature

**Comment by rumpl** [comment]
@kingdaro thanks!

This seems like a really big feature, I am not sure I will have time to take care of it soon, but we can always hope. Thanks for unstaling it, I can only promise I will try to see what I can do about it (currently I am no longer working in node world as you can guess by my github activity).

**Comment by sveyret** [comment]
@kingdaro, _depcheck_ should already manage workspaces since v0.9.1. What kind of problem do you have?

**Comment by benjie** [comment]
If you're using yarn2 / pnp then yarn's doctor script will tell you about missing dependencies:

```
yarn dlx @yarnpkg/doctor .
```

I'm not sure if it tells you about extraneous dependencies though.

We [use `lerna exec --stream 'yarn depcheck'` in Graphile Starter](https://github.com/graphile/starter/blob/e565f9927672f6825ee972e4ded216a6f325e521/package.json#L16), but it has a significant list of ignores.

**Comment by kingdaro** [comment]
@sveyret @benjie These are some good workarounds, though I can't upgrade to yarn 2, use pnpm, and I'd like to avoid using lerna as well, so the suggestions are unfortunately a no-go 😅 I'll probably figure something out if I need to, though.

**Comment by benjie** [comment]
Instead of `lerna exec --stream 'yarn depcheck'` you could do `for I in packages/*/; do cd $I; yarn depcheck; cd -; done`; though persisting the exit state would be more challenging and you can't leverage the parallelism this way.

**Comment by zraineri** [comment]
Inspired by @benjie 

For windows users not on yarn2/pnp  or using Lerna:
`FOR /D /r %G in ("packages\*") DO (cd packages/%~nxG & npx depcheck & cd../..)`

I just added a script named "depcheck" into my package.json and run with "yarn depcheck"

When I have a moment I'm going to put together a batch script which returns an easier to read report only showing packages which have dependency issues, but this works for now

**Comment by tvvignesh** [comment]
Hi. Just pitching in here. Is there a planned support for [PNPM Workspaces](https://pnpm.js.org/en/workspaces)? Would be great to have that. Thanks.

**Comment by sveyret** [comment]
@tvvignesh what kind of support do you need? As said [here](https://github.com/depcheck/depcheck/issues/245#issuecomment-632572647), usage of _depcheck_ with _pnpm_ is already possible…

**Comment by tvvignesh** [comment]
@sveyret My bad. Since I got a few false alerts like this and I saw this discussion I thought PNPM workspaces was not supported.

<img width="185" alt="1" src="https://user-images.githubusercontent.com/1165845/99570086-731eee80-29f7-11eb-916d-83ce314376f6.png">

Just did a test now. Looks like this may not be because of this issue with workspaces but a false alert by itself. All these packages are required by the packages within `node_modules` For eg: `@graphql-mesh/cli` which was installed in the root node_module of PNPM workspace which uses these deps but is detected as unused:

<img width="581" alt="2" src="https://user-images.githubusercontent.com/1165845/99570438-f0e2fa00-29f7-11eb-9b32-ab8c6050e16e.PNG">

Might not be related.

**Comment by Hypnosphi** [comment]
Here’s some activity

**Comment by eric-burel** [comment]
Ended up with the following command in my package.json that gives more clarity:
```json
    "depcheck": "lerna exec --no-bail -- 'echo \"\n***** $LERNA_PACKAGE_NAME *****\" && depcheck'"
```

Note that `yarn run lerna exec` won't work as expected with the `--no-bail` option (this messes up the option order, because you run a command with yarn that runs a command with Lerna...). If you need to quick test this command instead do `./node_modules/.bin/lerna exec --no-bail -- depcheck`.

**Comment by clemsos** [comment]
`yarn workspace foreach dlx depcheck` seems to work okay with yarn@2.4.2

**Comment by olof-nord** [comment]
As someone who want to introduce npm workspaces for a monorepo, I am wondering if depcheck at all works.

As dependencies used in more than one sub-project are moved (hoisted?) to the root node_modules, basically several `dependencies` are reported as unused.

I imagine if the sub-project package.json in combination with the local as well as root node_modules would be taken into account, the dependencies would (could?) be correctly detected.

**Comment by jonaskello** [comment]
If anyone is looking for a bash workaround this oneliner in as a package.json script worked out quite well for me:

```
    "depcheck-all": "for d in packages/*; do (cd \"$d\" && echo && echo -n '-----> DIR: ' &&pwd && yarn depcheck); done"
```

**Comment by robdefeo** [comment]
Commenting to un-stale issue. It's an important feature, we and others use monorepo's and this would be helpful as it's easy to add dependencies but checking if they are no longer needed is more problematic.

**Comment by bblatnick32** [comment]
Commenting to un-stale.  As anyone found a workaround for a lerna monorepo? Bascially checking the global node_modules as well as the package node_modules in conjuction for each package?

**Comment by webpro** [comment]
First of all, apologies for my impolite entry. But seeing that this project doesn't receive many updates (the latest release is released almost a year ago?) and @rumpl seems to be be [occupied with other things](https://github.com/depcheck/depcheck/issues/245#issuecomment-632345110), I would like to inform there is an alternative tool available that supports workspaces: Knip. Disclaimer: I'm the creator. I didn't develop a drop-in replacement for depcheck (I started the project to find unused exports), but in terms of features and "specials" there is quite some overlap in the upcoming v1 release. The command to do roughly the same thing as depcheck would be:

```
knip --include dependencies,unlisted
```

You can find the latest version at https://github.com/webpro/knip/tree/next. I'm linking to the `next` version, since the latest `dist` version (v0.13.3) does not support workspaces. To install this:

```
npm install -D knip@next
```

Please note that Knip needs a little bit of configuration, it may have bugs and will probably give false positives, but at the same time I guess users of tools like this understand. Happy to hear any feedback, questions, bug reports, etcetera (at the Knip repo, not here obviously).

This is open source, and maybe Knip can take some of the load of @rumpl's shoulders, who did an absolutely great job creating depcheck and helping tons of people and projects. Obviously I took some inspiration from this project and I'm grateful for that. Cheers!

**Comment by willbamford** [comment]
This works for my Yarn 4 monorepo:
```
yarn workspaces foreach -A dlx depcheck
```
